gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/NewDagManager.java

```diff
@@ -0,0 +1,429 @@
+    addDag(dag, persist, setStatus);
+    // Only the active newDagManager should delete the flowSpec
+    if (isActive) {
+      deleteSpecFromCatalogIfAdhoc(flowSpec);
+    }
+  }
+
+  private synchronized void addDag(Dag<JobExecutionPlan> dag, boolean persist, boolean setStatus) throws IOException {
+    // TODO: Used to track missing dag issue, remove later as needed
+    log.info("Add dag (persist: {}, setStatus: {}): {}", persist, setStatus, dag);
+    if (!isActive) {
+      log.warn("Skipping add dag because this instance of DagManager is not active for dag: {}", dag);
+      return;
+    }
+
+    DagManager.DagId dagId = DagManagerUtils.generateDagId(dag);
+    String dagIdString = dagId.toString();
+    if (persist) {
+      // Persist the dag
+      this.dagStateStore.writeCheckpoint(dag);
+      // After persisting the dag, its status will be tracked by active dagManagers so the action should be deleted
+      // to avoid duplicate executions upon leadership change
+      if (this.dagActionStore.isPresent()) {
+        this.dagActionStore.get().deleteDagAction(dagId.toDagAction(DagActionStore.FlowActionType.LAUNCH));
+      }
+    }
+
+    DagActionStore.DagAction dagAction = new DagActionStore.DagAction(
+        dagId.getFlowGroup(), dagId.getFlowName(), dagId.getFlowExecutionId(), DagActionStore.FlowActionType.LAUNCH);
+
+    if (dagManagementStateStore.containsDag(dagIdString)) {
+      log.warn("Already tracking a dag with dagId {}, skipping.", dagIdString);
+      return;
+    }
+
+    dagManagementStateStore.addDag(dagIdString, dag);
+    this.dagTaskStream.addDagAction(dagAction);
+
+    if (setStatus) {
+      DagManagerUtils.submitPendingExecStatus(dag, this.eventSubmitter);
+    }
+  }
+
+  private void deleteSpecFromCatalogIfAdhoc(FlowSpec flowSpec) {
+    if (!flowSpec.isScheduled()) {
+      this.flowCatalog.remove(flowSpec.getUri(), new Properties(), false);
+    }
+  }
+
+  public void removeDagActionFromStore(DagActionStore.DagAction dagAction) throws IOException {
+    if (this.dagActionStore.isPresent()) {
+      this.dagActionStore.get().deleteDagAction(dagAction);
+    }
+  }
+
+  //@Subscribe todo uncomment it when new dag manager is ready for kill functionality
+  public void handleKillFlowEvent(KillFlowEvent killFlowEvent) {
+    String flowGroup = killFlowEvent.getFlowGroup();
+    String flowName = killFlowEvent.getFlowName();
+    if (killFlowEvent.getFlowExecutionId() != null) {
+      long flowExecutionId = killFlowEvent.getFlowExecutionId();
+      log.info("Received kill request for flow ({}, {}, {})", flowGroup, flowName, flowExecutionId);
+      // todo - should this be added to the dac action store instead?
+      this.dagTaskStream.addDagAction(new DagActionStore.DagAction(flowGroup, flowName, String.valueOf(flowExecutionId), DagActionStore.FlowActionType.KILL));
+
+    } else {
+      List<Long> flowExecutionIds = getJobStatusRetriever().getLatestExecutionIdsForFlow(flowName, flowGroup, 10);
```

Comment:
one note here that 10 number is hard coded and somewhat arbitrary although I agree it's 'enough' to kill all remaining flows in most cases. we can make it configurable or add a comment to note this

Reply:
probably in a separate PR